lib/plugins/chat.js

```diff
@@ -110,14 +110,16 @@ function inject (bot, options) {
       }
     }
 
-    bot._client.once('tab_complete', (packet) => {
-      cb(undefined, packet.matches)
-    })
+    return new Promise(resolve => {
+      bot._client.once('tab_complete', (packet) => {
+        resolve(packet.matches)
+      })
 
-    bot._client.write('tab_complete', {
-      text,
-      assumeCommand,
-      lookedAtBlock: position
+      bot._client.write('tab_complete', {
+        text,
+        assumeCommand,
+        lookedAtBlock: position
+      })
```

Comment:
```js
bot._client.write('tab_complete', {
  text,
  assumeCommand,
  lookedAtBlock: position,
})

const [{ matches }] = await once(bot._client, 'tab_complete')
return matches
```

Reply:
That doesn't work, because the listener needs to be registered _before_ triggering it otherwise you'll introduce a race condition. But if you put it after, it will await infinitely because it never gets triggered.

Reply:
You can't really have a race condition, since having a race condition would mean that you have send & received the packet in the same event loop turn, it's impossible but if you are really bothered by that you can just do

```js
const response = once(bot._client, 'tab_complete')

bot._client.write('tab_complete', {
  text,
  assumeCommand,
  lookedAtBlock: position,
})

const [{ matches }] = await response
return matches
```

Reply:
That is better, thanks. I never want to assume stuff like that, I'm a bit paranoid about this. Look at `inventory#clickWindow` for example, it definitely requires this structure.

Reply:
I understand, but the way nodejs works you can't have this kind of race condition, but since not everyone is aware of how it works it might be better to write the code without taking this into account so everyone agrees on the code

Reply:
I know, but what if later on that code is changed to a shortcut version or something that directly emits the event, I think it will make it fail, right? Or am I wrong in this assumption?

Reply:
Yep in this case it will fail since emit is synchronous but since it we are dealing with protocol I think it is safe, but yeah it might not work for other cases

Reply:
resolved in 43ba7aa